features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularNetwork.cpp

```diff
@@ -0,0 +1,344 @@
+             success && (protocol <= nsapi_security_to_modem_security(NSAPI_SECURITY_CHAP)); protocol++) {
+            if ((_auth == NSAPI_SECURITY_UNKNOWN) || (nsapi_security_to_modem_security(_auth) == protocol)) {
+                _at.cmd_start("AT+UPSD=0,6,");
+                _at.write_int(protocol);
+                _at.cmd_stop();
+                _at.resp_start();
+                _at.resp_stop();
+
+                if (_at.get_last_error() == NSAPI_ERROR_OK) {
+                    // Activate, wait upto 30 seconds for the connection to be made
+                    _at.set_at_timeout(30000);
+                    _at.cmd_start("AT+UPSDA=0,3");
+                    _at.cmd_stop();
+                    _at.resp_start();
+                    _at.resp_stop();
+                    _at.restore_at_timeout();
+
+                    if (_at.get_last_error() == NSAPI_ERROR_OK) {
+                        activated = true;
+                    }
+                }
+            }
+        }
+    }
+
+    return activated;
+}
+
+// Convert nsapi_security_t to the modem security numbers
+int UBLOX_AT_CellularNetwork::nsapi_security_to_modem_security(nsapi_security_t nsapi_security)
+{
+    int modem_security = 3;
+
+    switch (nsapi_security)
+    {
+        case NSAPI_SECURITY_NONE:
+            modem_security = 0;
+            break;
+        case NSAPI_SECURITY_PAP:
+            modem_security = 1;
+            break;
+        case NSAPI_SECURITY_CHAP:
+            modem_security = 2;
+            break;
+        case NSAPI_SECURITY_UNKNOWN:
+            modem_security = 3;
+            break;
+        default:
+            modem_security = 3;
+            break;
+    }
+
+    return modem_security;
+}
+
+// Disconnect the on board IP stack of the modem.
+bool UBLOX_AT_CellularNetwork::disconnect_modem_stack()
+{
+    bool success = false;
+
+    if (get_ip_address() != NULL) {
+        _at.cmd_start("AT+UPSDA=" PROFILE ",4");
+        _at.cmd_stop();
+        _at.resp_start();
+        _at.resp_stop();
+
+        if (_at.get_last_error() == NSAPI_ERROR_OK) {
+            success = true;
+        }
+    }
+
+    return success;
+}
+
+nsapi_error_t UBLOX_AT_CellularNetwork::get_imsi(char* imsi)
+{
+    _at.lock();
+    _at.cmd_start("AT+CIMI");
+    _at.cmd_stop();
+    _at.resp_start();
+    int len = _at.read_string(imsi, MAX_IMSI_LENGTH);
+    if (len > 0) {
+        imsi[len] = '\0';
+    }
+    _at.resp_stop();
+
+    return _at.unlock_return_error();
+}
+
+// Get the next set of credentials, based on IMSI.
+void UBLOX_AT_CellularNetwork::get_next_credentials(char ** config)
+{
+    if (*config) {
+        _apn    = _APN_GET(*config);
+        _uname  = _APN_GET(*config);
+        _pwd    = _APN_GET(*config);
+    }
+}
+
+const char *UBLOX_AT_CellularNetwork::get_gateway()
+{
+    return get_ip_address();
+}
+
+nsapi_error_t UBLOX_AT_CellularNetwork::detach()
+{
+    _at.lock();
```

Comment:
After detach we are receiving two urc's, '+CGREG' and '+UUPSDD', to process these urc's we have added 50 ms wait here otherwise "CellularNetwork test detach" fails.

Reply:
Could you add wait just before when "test detach" is called?

Reply:
Are you suggesting to add wait in `CellularNetwork test detach` test case?
if yes then I can remove my `detach()` function and will use parent implementation.